In [1]:
#imports
import pandas as pd
%matplotlib inline 
import matplotlib.pyplot as plt
from matplotlib import dates as mpl_dates
from matplotlib.pyplot import figure
import numpy as np
import seaborn as sns
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()
import calendar

In [2]:
df_merchandise = pd.read_excel('Merchandise.xlsx', index_col="ProductID")

## Data Preprocessing & Analysis

In [3]:
df_merchandise.head()

ProductType ProductColor ProductSize        ProductDescription  \
ProductID                                                                  
-2               None         None        None                      Lamp   
-1               None         None        None  Wrought Iron W. 5 Chairs   
 0             filler       filler      filler                    filler   
 1               None         None        None             Candle Holder   
 2               None         None        None                     Colum   

           OrigPrice   Price  EnterDate TransferDate  TransferPrice  \
ProductID                                                             
-2            147.60  147.60 2018-08-18   2018-08-18         147.60   
-1            124.75  124.75 2018-07-03   2018-07-03         124.75   
 0               NaN    0.00 1999-12-31   2018-04-14          68.00   
 1              2.00    2.00 2020-05-20   2020-05-20           2.00   
 2             49.50   49.50 2020-07-18   2020-07-18          49.50   

          TransferInd  ... ListOnline  CustNo  FixSplit  TaxRate  PhyDate  \
ProductID              ...                                                  
-2               Sold  ...      False     NaN     False      NaN      NaN   
-1               Sold  ...      False     NaN     False      NaN      NaN   
 0               Sold  ...      False     NaN     False      NaN      NaN   
 1               Sold  ...      False     NaN     False      NaN      NaN   
 2               Sold  ...      False     NaN     False      NaN      NaN   

          Auction2 ConCode BuyOut  CardFee  Reminded  
ProductID                                             
-2             NaN     NaN  False      0.0     False  
-1             NaN     NaN  False      0.0     False  
 0             NaN     NaN  False      0.0     False  
 1             NaN     NaN  False      0.0     False  
 2             NaN     NaN  False      0.0     False  

[5 rows x 50 columns]

In [4]:
df_merchandise.shape

(15308, 50)

In [5]:
df_merchandise.count()

ProductType           15308
ProductColor          15308
ProductSize           15308
ProductDescription    15308
OrigPrice             15307
Price                 15308
EnterDate             15308
TransferDate          10923
TransferPrice         15308
TransferInd           15308
TransferBy            15308
CustomerID            15308
ReqPrice              15308
InvoiceNumber         10906
Return                15308
MarkdownDate1          9976
MarkdownDate2          8497
MarkdownDate3          6976
LabelPrinted          15308
MercCut               15307
DownPymt              15308
DownWho                   0
DownDate                  0
ItemCount             15308
Date1                     0
Text1                   395
Num1                  15308
ExpireDate            15308
LayawayGrp                0
PCName                14925
DontPrintTag          15308
ItemNotes                 0
WholesalePrice        15308
BuyersFee             15308
DownPhone                 0
ResaleNow           

In [6]:
df_merchandise['DownPymt'].unique()
# will drop

array([0])

In [7]:
df_merchandise['Num1'].unique()
# will drop

array([0])

In [8]:
df_merchandise['Taxable'].unique()

array([nan, 'Y', 'N'], dtype=object)

In [9]:
df_merchandise['Taxable'].value_counts()
# will drop

Y    14910
N       15
Name: Taxable, dtype: int64

In [10]:
df_merchandise['NeverMD'].unique()

array([nan, 'Y', 'N'], dtype=object)

In [11]:
df_merchandise['NeverMD'].value_counts()
# will drop, unable to find documentation for this column description

N    13159
Y     1766
Name: NeverMD, dtype: int64

In [12]:
df_merchandise['TaxRate'].unique()
# will drop, these values will be worthless

array([nan,  0.])

In [13]:
df_merchandise['CardFee'].unique()
# will drop, these values will be worthless

array([ 0., nan])

In [14]:
# drop columns with no/many missing values and ones that are unnecessary
df_merchandise = df_merchandise.drop(['DownWho','DownDate','Date1','Text1','LayawayGrp','ItemNotes','Num1',
                                      'DownPhone','PreviousOwner','LabelPrinted','PCName','DontPrintTag',
                                      'Taxable','WholesalePrice','AuctionID','CustNo','PhyDate','Auction2',
                                      'DownPymt','ResaleNow','ListOnline','FixSplit','ConCode','NeverMD',
                                      'BuyOut','TaxRate','Reminded','CardFee'], axis=1)

In [15]:
# drop filler row
df_merchandise = df_merchandise.drop([0])
df_merchandise.head()

ProductType ProductColor ProductSize        ProductDescription  \
ProductID                                                                  
-2               None         None        None                      Lamp   
-1               None         None        None  Wrought Iron W. 5 Chairs   
 1               None         None        None             Candle Holder   
 2               None         None        None                     Colum   
 3               None         None        None             Credenza Lamp   

           OrigPrice   Price  EnterDate TransferDate  TransferPrice  \
ProductID                                                             
-2            147.60  147.60 2018-08-18   2018-08-18         147.60   
-1            124.75  124.75 2018-07-03   2018-07-03         124.75   
 1              2.00    2.00 2020-05-20   2020-05-20           2.00   
 2             49.50   49.50 2020-07-18   2020-07-18          49.50   
 3             71.10   71.10 2018-08-25   2018-08-25          71.10   

          TransferInd  ... ReqPrice  InvoiceNumber  Return  MarkdownDate1  \
ProductID              ...                                                  
-2               Sold  ...      0.0         8150.0   False            NaT   
-1               Sold  ...      0.0         7749.0   False            NaT   
 1               Sold  ...      0.0        13211.0   False            NaT   
 2               Sold  ...      0.0        13842.0   False            NaT   
 3               Sold  ...      0.0         8225.0   False            NaT   

           MarkdownDate2 MarkdownDate3 MercCut ItemCount  ExpireDate  \
ProductID                                                              
-2                   NaT           NaT     0.0         0  2018-11-16   
-1                   NaT           NaT     0.5         0  2018-10-01   
 1                   NaT           NaT     0.0         0  2020-08-18   
 2                   NaT           NaT     0.5         0  2020-10-16   
 3                   NaT           NaT     0.0         0  2018-11-23   

           BuyersFee  
ProductID             
-2               0.0  
-1               0.0  
 1               0.0  
 2               0.0  
 3               0.0  

[5 rows x 22 columns]

In [16]:
df_merchandise.dtypes

ProductType                   object
ProductColor                  object
ProductSize                   object
ProductDescription            object
OrigPrice                    float64
Price                        float64
EnterDate             datetime64[ns]
TransferDate          datetime64[ns]
TransferPrice                float64
TransferInd                   object
TransferBy                    object
CustomerID                     int64
ReqPrice                     float64
InvoiceNumber                float64
Return                          bool
MarkdownDate1         datetime64[ns]
MarkdownDate2         datetime64[ns]
MarkdownDate3         datetime64[ns]
MercCut                      float64
ItemCount                      int64
ExpireDate            datetime64[ns]
BuyersFee                    float64
dtype: object

In [17]:
# MercCut signifies 0.5 for items consigned (50/50 split), 0.0 for items the store purchased new
df_merchandise.MercCut.value_counts()

# 13,255 of the items have been consigned

0.5    13255
0.0     2052
Name: MercCut, dtype: int64

In [18]:
# Convert invoice number to int from float
df_merchandise.InvoiceNumber = df_merchandise.InvoiceNumber.fillna(0.0).astype(int)
df_merchandise.InvoiceNumber.dtype

dtype('int64')

In [19]:
df_merchandise.shape

(15307, 22)

In [20]:
# View top 21 selling items and their value counts
df_merchandise['ProductType'].value_counts().head(21)

Accessory       5266
Chair           1194
Decor            655
Table            402
None             384
Pillow           371
Art              365
Sofa             296
Rug              293
End Table        228
Lamp             224
Picture          221
Desk             171
Light            155
Bed              140
Barstool         124
Mirror           113
Coffee Table     111
Chest            104
Dining Table     104
Console           99
Name: ProductType, dtype: int64

In [21]:
products = df_merchandise['ProductType'].value_counts()
products[products < 10].shape
# there are 1103 different entry labels for the Product Type column that occur 10 or less times 
# I will likely exclude these entries from the dataframe when analyzing products since cleaning 1000+ rows is
# not feasible given the timeline of this project

(1103,)

In [22]:
products[products < 100].shape
# Outside of the top 20 selling items listed in ProductType (#20 having 104 entries), there are 1191 different 
# entry labels for this category. Only the top 20 items for this category will be analyzed

(1191,)

In [23]:
df_merchandise[df_merchandise.groupby('ProductType')['ProductType'].transform('size') < 100].shape
# 4,386 rows will be excluded from top20 item analysis prior to cleaning

(4386, 22)

In [24]:
df_merchandise['ProductColor'].value_counts().head(20)

None        14779
White          53
Brown          44
Black          29
Ivory          25
Flint          22
Driftwoo       19
Silver         17
Linen          13
Seaglass       12
Wheat          12
Sapphire       12
Gold           11
Blue           11
Red            10
Gray            9
Wood            9
Teal            9
Multi           8
Navy            8
Name: ProductColor, dtype: int64

Since there's only 528 entires where the Product Color is any description other than "None", I will remove this column from the dataset as it will not be useful for analysis. 

In [25]:
df_merchandise = df_merchandise.drop(['ProductColor'], axis=1)

In [26]:
df_merchandise['ProductSize'].value_counts().head(20)

None          14798
Jubilee          94
Park Avenu       39
Queen            18
Small            18
Medium           17
King             14
Large            13
Pomeroy          11
53006225         10
B446              8
CM1700            8
CM4550            8
CM3500            8
53005996          7
CM3200            7
Wood              6
9ft X 4ft         6
22x22             5
53005413el        3
Name: ProductSize, dtype: int64

Only 509 of the rows contain useful information regarding Product Size (other than 'None), will drop this column

In [27]:
df_merchandise = df_merchandise.drop(['ProductSize'], axis=1)

In [28]:
df_merchandise['ProductDescription'].value_counts().head(20)

Pillow          611
Lamp            436
Picture         347
Art             315
Chair           296
Rug             228
Mirror          216
Vase            203
Decor           202
Chandelier      145
Barstool        108
Free             98
Sofa             95
Book             88
Desk             87
Nightstand       78
Sidetable        69
Plate            68
Office Chair     65
Finial           65
Name: ProductDescription, dtype: int64

In [29]:
#How many rows are there with a Product Description label that occurs less than 65 times (outside of top20 values)
df_merchandise[df_merchandise.groupby('ProductDescription')['ProductDescription'].transform('size') < 65].shape

(11487, 20)

There are too many rows to go through and clean for Product Description given the timeframe of this report. However, there are rows in which the value for 'ProductType' contains 'None' or 'Accessory', but has a relevant value for 'ProductDescription' that could replace the ProductType value and fit into a pre-existing label in ProductType such as 'Table','Chair', etc.

In [30]:
accessories = df_merchandise[df_merchandise.ProductType == 'Accessory']
accessories.ProductDescription.value_counts().head(15)

Pillow             313
Lamp               269
Picture            226
Art                220
Vase               142
Mirror             129
Decor              118
Rug                 80
Free                72
Chandelier          72
Pewter Figurine     53
Plate               46
Painting            39
Book                37
Finial              36
Name: ProductDescription, dtype: int64

In [31]:
df_merchandise[df_merchandise.ProductType == 'Rug'].shape
# It looks like there are sufficient values for PD under PT = 'Accessory' that are already in the top 20
# value counts for PT that can be replaced using their PD values 

(293, 20)

In [32]:
# Add filter to main replacement cell below
table = df_merchandise[df_merchandise.ProductType.str.contains('table', case=False)]
table.ProductType.value_counts()

Table              402
End Table          228
Coffee Table       111
Dining Table       104
Sidetable           79
Side Table          48
Endtable            44
Sofatable           38
Coffeetable         29
Diningtable         23
Sofa Table          17
Round Table          8
Table Set            8
Nesting Tables       6
Diningtable Set      5
Coffetable           4
Table/6 Chairs       4
Entry Table          4
Pub Table            3
Table 8 Chairs       3
Table Lamp           3
Dining Table W/      3
Table 6 Chairs       3
Table 4 Chairs       3
Cocktail Table       3
Diningtable W/8      3
Dining Table Se      2
Dinning Table        2
Stacking Table       2
TABLE                2
                  ... 
Table W/7Chair       1
Table And CHair      1
Table/4 Chairs       1
DRAFTING TABLE       1
Red Table            1
Table W/chairs       1
Parquet Table        1
Chairs/table         1
Half Moon Table      1
Table/Chairs         1
Dinging Table        1
Adjustable Base      1
End Tables 

In [33]:
# Add filter to main replacement cell below
chairs = df_merchandise[df_merchandise.ProductType.str.contains('chair', case=False)]
chairs.ProductType.value_counts()

Chair              1194
Chairs               34
Dining Chair         12
Chair/ottoman        10
Chair W/ottoman       7
Chair And Ottom       6
Dinning Chair         6
Chair W/ott           6
Barrel Chair          4
Desk Chair            4
Rocking Chair         4
Metal Chair           4
Table/6 Chairs        4
COPPER CHAIR          4
Office Chair          4
Table 6 Chairs        3
4 Chairs              3
DINING CHAIR          3
Kai Side Chair        3
6 Chairs              3
Table 8 Chairs        3
Table 4 Chairs        3
Henredon Chair        2
Side Chair            2
Swivel Chair          2
Antique Chair         2
Formal Chair          2
Parson Chair          2
Sherrill Chair        2
Arm Chair             2
                   ... 
Table/4 Chairs        1
Chair & Ottoman       1
Table/chairs          1
Set Of 4 Chairs       1
Chairside             1
Table 7/chairs        1
Chairs Set/4          1
Bernhardt Chair       1
RH 2 Chairs/ Ot       1
Leather Chairs        1
Chair W/green S 

In [34]:
# Replace ProductType values for 'Accessory' that also contain top 20 values of ProductType for the ProductDescription category
# Replace values for 'Pillow'
pillow_filter = (df_merchandise['ProductType'] == 'Accessory') & (df_merchandise['ProductDescription'] == 'Pillow')
df_merchandise.loc[pillow_filter, 'ProductType'] = df_merchandise['ProductDescription']

In [35]:
# Ensure new values were correctly added to 'Pillow' value counts and deducted from 'Accessory'
# 'Pillow' should increase from 371 to 684
df_merchandise.ProductType.value_counts().head()

Accessory    4953
Chair        1194
Pillow        684
Decor         655
Table         402
Name: ProductType, dtype: int64

In [36]:
# Ensure this swap did not affect ProductDescription value counts for 'Pillow'
accessories.ProductDescription.value_counts().head()

Pillow     313
Lamp       269
Picture    226
Art        220
Vase       142
Name: ProductDescription, dtype: int64

In [37]:
# Repeat for chairs
chairs = df_merchandise[(df_merchandise['ProductType'] == 'Accessory') & (df_merchandise['ProductDescription'] == 'Chair')]
chairs.shape

(10, 20)

In [38]:
# Repeat for chairs
chair_filter = (df_merchandise['ProductType'] == 'Accessory') & (df_merchandise['ProductDescription'] == 'Chair')
df_merchandise.loc[chair_filter, 'ProductType'] = df_merchandise['ProductDescription']

In [39]:
# Ensure new values were correctly added to 'Pillow' value counts and deducted from 'Accessory'
# 'Chair' value should increase from 1194 to 1204
df_merchandise.ProductType.value_counts().head(10)

Accessory    4943
Chair        1204
Pillow        684
Decor         655
Table         402
None          384
Art           365
Sofa          296
Rug           293
End Table     228
Name: ProductType, dtype: int64

Repeat replacement process for other ProductType categories to better distribute values from **both** 'Accessory' and 'None' in ProductTypw that would better fit into other categories.

Check 'None' values for items over $200 to correct for

In [52]:
none_filter = df_merchandise[(df_merchandise['ProductType'] == 'None') & ((df_merchandise.Price >200))]
none_filter

ProductType   ProductDescription  OrigPrice     Price  EnterDate  \
ProductID                                                                    
19               None                Bench     500.00   500.000 2020-07-25   
18812            None      RH Stacking Bin     371.25   259.875 2018-06-02   
20921            None              Dresser     431.20   431.200 2018-09-01   
20957            None  Grey/olive Loveseat     237.00   237.000 2018-09-07   
21335            None   2 Twin Foundations     270.00   270.000 2018-10-02   
22230            None     Leather Recliner    1000.00  1000.000 2018-11-24   
23057            None  Window Pane Divider     300.00   300.000 2019-01-19   
23598            None             Shutters     375.00   375.000 2019-02-18   
23730            None     King Mat And Box     400.00   400.000 2019-03-04   
24235            None      3 Twin/ 1 Queen     250.00   250.000 2019-03-30   
24241            None                Hutch     350.00   350.000 2019-03-30   
24638            None            * Unknown     750.00   750.000 2019-04-19   
24640            None        Queen Mat/box     389.00   389.000 2019-04-19   
25037            None                 None     750.00   502.500 2019-05-04   
26541            None           Q Matt/box     450.00   450.000 2019-07-09   
27242            None  Hollywood Sectional    1098.00   768.600 2019-08-28   
27306            None    Barn Doors 2 Sets     900.00   900.000 2019-08-31   
27374            None            * Unknown     400.00   400.000 2019-09-10   
27376            None            * Unknown    2400.00  2400.000 2019-09-10   
27616            None               Buffet     450.00   450.000 2019-10-02   
28565            None             Cow Hide     495.00   495.000 2019-11-23   
29013            None            * Unknown     216.00   216.000 2019-12-16   
29780            None            * Unknown     225.00   225.000 2020-01-20   
29944            None            * Unknown    1098.00  1098.000 2020-01-28   
30181            None            * Unknown     800.00   800.000 2020-02-09   
30220            None                Glass     275.00   275.000 2020-02-13   
31506            None       Bubble Ligting     425.00   425.000 2020-05-21   
31572            None            Universal     412.50   412.500 2020-05-23   
32160            None               Mantle     350.00   350.000 2020-06-18   
32246            None         Leaning Door     349.30   349.300 2020-06-20   
32986            None            * Unknown     550.00   495.000 2020-07-11   
33199            None         New Mattress     700.00   700.000 2020-07-27   
55658            None            * Unknown     425.00   425.000 2019-01-07   
26363282         None        Black Pendant     291.00   291.000 2019-02-09   
26363283         None            * Unknown     291.00   291.000 2019-02-09   

          TransferDate  TransferPrice TransferInd TransferBy  CustomerID  \
ProductID                                                                  
19          2020-07-25         500.00        Sold       Noel         237   
18812       2018-06-05           0.00   Available     System         237   
20921       2018-09-01         431.20        Sold    Adriana           1   
20957       2018-09-07         237.00        Sold    Adriana           1   
21335       2018-10-02         270.00        Sold       Noel           1   
22230       2018-11-24        1000.00        Sold        Ken           1   
23057       2019-01-19         300.00        Sold    Adriana           1   
23598       2019-02-18         375.00        Sold    Adriana           1   
23730       2019-03-04         400.00        Sold        Ken           1   
24235       2019-03-30         250.00        Sold    Adriana           1   
24241       2019-03-30         350.00        Sold    Adriana        1677   
24638       2019-04-19         750.00        Sold       Gina           1   
24640       2019-04-19         389.00        Sold        Ke

In [40]:
# Accessory and None values cleaning

# Replace values where 'lamp' occurs
lamp_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('lamp', case=False))
df_merchandise.loc[lamp_filter, 'ProductType'] = 'Lamp'

# Repeat for chairs
chair_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('chair', case=False))
df_merchandise.loc[chair_filter, 'ProductType'] = 'Chairs'

# Repeat for pillows where 'pillow' occurs
pillow_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('pillow', case=False))
df_merchandise.loc[pillow_filter, 'ProductType'] = 'Pillow'

# Repeat for tables
table_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & ((df_merchandise['ProductDescription'].str.contains('table', case=False)) & ~(df_merchandise['ProductDescription'].str.contains('unforgettable', case=False)))
df_merchandise.loc[table_filter, 'ProductType'] = 'Table'

# Repeat for art
art_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('art', case=False))
df_merchandise.loc[art_filter, 'ProductType'] = 'Art'

# Repeat for sofa
sofa_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('sofa', case=False))
df_merchandise.loc[sofa_filter, 'ProductType'] = 'Sofa'

# Repeat for rug
rug_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('rug', case=False))
df_merchandise.loc[rug_filter, 'ProductType'] = 'Rug'

# Repeat for pictures as 'Art'
picture_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('picture', case=False))
df_merchandise.loc[picture_filter, 'ProductType'] = 'Art'

# Repeat for desk
desk_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('desk', case=False))
df_merchandise.loc[desk_filter, 'ProductType'] = 'Desk'

# Repeat for bed
bed_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('bed', case=False))
df_merchandise.loc[bed_filter, 'ProductType'] = 'Bed'

# Repeat for chest
chest_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('chest', case=False))
df_merchandise.loc[chest_filter, 'ProductType'] = 'Chest'

# Repeat for mirror
mirror_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('mirror', case=False))
df_merchandise.loc[mirror_filter, 'ProductType'] = 'Mirror'

# Repeat for painting as 'Art'
painting_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (df_merchandise['ProductDescription'].str.contains('painting', case=False))
df_merchandise.loc[painting_filter, 'ProductType'] = 'Art'

# Repeat for lights
light_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & ((df_merchandise['ProductDescription'].str.contains('light', case=False)) & ~(df_merchandise['ProductDescription'].str.contains('sconce', case=False)))
df_merchandise.loc[light_filter, 'ProductType'] = 'Light'

# Repeat for chandelier, lantern, 'ligt' as 'Lighting'
chandelier_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (((df_merchandise['ProductDescription'].str.contains('chandelier', case=False)))
                                                                      | ((df_merchandise['ProductDescription'].str.contains('lantern', case=False))) | | ((df_merchandise['ProductDescription'].str.contains('ligt', case=False)))
df_merchandise.loc[chandelier_filter, 'ProductType'] = 'Lighting'

# Repeat for clock, vase, book, pewter items, bowl, greenery, tray, cross, urn, plate, decor as 'Decor'
decor_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & (((df_merchandise['ProductDescription'].str.contains('decor', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('clock', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('book', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('pewter', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('bowl', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('greenery', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('tray', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('cross', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('urn', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('plate', case=False)))
                                                                | ((df_merchandise['ProductDescription'].str.contains('vase', case=False))))
df_merchandise.loc[decor_filter, 'ProductType'] = 'Decor'

# Repeat for free
free_filter = ((df_merchandise['ProductType'] == 'Accessory') | (df_merchandise['ProductType'] == 'None')) & ((df_merchandise['ProductDescription'].str.contains('free', case=False)))
df_merchandise.loc[free_filter, 'ProductType'] = 'Free'

# Repeat for sleeper as 'Sofa'
none_sleeper_filter = (df_merchandise['ProductType'] == 'None') & ((df_merchandise['ProductDescription'].str.contains('sleeper', case=False)))
df_merchandise.loc[none_sleeper_filter, 'ProductType'] = 'Sofa'

# Repeat for hardware, finials, rods as 'Accessory'
none_hw_filter = (df_merchandise['ProductType'] == 'None') & ((df_merchandise['ProductDescription'].str.contains('hardware', case=False))
                                                             |(df_merchandise['ProductDescription'].str.contains('finial', case=False))
                                                             |(df_merchandise['ProductDescription'].str.contains('rod', case=False)))
df_merchandise.loc[none_hw_filter, 'ProductType'] = 'Accessory'

# Repeat for 'mat' values indicating mattresses as 'Bed'
none_mattress_filter = (df_merchandise['ProductType'] == 'None') & ((df_merchandise['ProductDescription'].str.contains('mat', case=False)))
df_merchandise.loc[none_mattress_filter, 'ProductType'] = 'Bed'

# Repeat for 'bench' values
none_bench_filter = (df_merchandise['ProductType'] == 'None') & ((df_merchandise['ProductDescription'].str.contains('bench', case=False)))
df_merchandise.loc[none_bench_filter, 'ProductType'] = 'Bench'

# Repeat for 'dresser' values
none_dresser_filter = (df_merchandise['ProductType'] == 'None') & ((df_merchandise['ProductDescription'].str.contains('bench', case=False)))
df_merchandise.loc[none_dresser_filter, 'ProductType'] = 'Dresser'

In [41]:
# Replace and consolidate all values in ProductType column

# Repeat for end table, coffee table, dining table, sidetable, etc. as 'Table' to consolidate
mass_table_filter = df_merchandise.ProductType.str.contains('table', case=False)
df_merchandise.loc[mass_table_filter, 'ProductType'] = 'Table'

# Repeat for lamp, light, chandelier as 'Lighting' to consolidate
lighting_filter = (df_merchandise['ProductType'] == 'Lamp') | (df_merchandise['ProductType'] == 'Light') | (df_merchandise['ProductType'] == 'LIghting') | (df_merchandise['ProductType'] == 'Chandelier')
df_merchandise.loc[lighting_filter, 'ProductType'] = 'Lighting'

# Repeat for all values of 'bed' into 'Bed' category
all_bed_filter = df_merchandise.ProductType.str.contains('bed', case=False)
df_merchandise.loc[all_bed_filter, 'ProductType'] = 'Bed'

# Repeat for all values of 'chair','barstool' into 'Chair' category in PT to consolidate
# contains 'dining chairs', 'sofa chair', 'side chair', etc.
mass_chair_filter = (df_merchandise.ProductType.str.contains('chair', case=False)) | (df_merchandise.ProductType.str.contains('barstool', case=False))
df_merchandise.loc[mass_chair_filter, 'ProductType'] = 'Chairs'

# Repeat for ottoman values 
ottoman_filter = df_merchandise.ProductType.str.contains('ottoman', case=False)
df_merchandise.loc[ottoman_filter, 'ProductType'] = 'Ottoman'

# Repeat for 'art' values 
mass_art_filter = df_merchandise.ProductType.str.contains('art', case=False)
df_merchandise.loc[mass_art_filter, 'ProductType'] = 'Art'

# Repeat for 'pillow' values 
mass_pillow_filter = df_merchandise.ProductType.str.contains('pillow', case=False)
df_merchandise.loc[mass_pillow_filter, 'ProductType'] = 'Pillow'

# Repeat for 'rug' values 
mass_rug_filter = df_merchandise.ProductType.str.contains('rug', case=False)
df_merchandise.loc[mass_rug_filter, 'ProductType'] = 'Rug'

# Repeat for remaining 'light' values not captured
mass_light_filter = df_merchandise.ProductType.str.contains('light', case=False)
df_merchandise.loc[mass_light_filter, 'ProductType'] = 'Lighting'

# Repeat for remaining 'sofa' values not captured
mass_sofa_filter = df_merchandise.ProductType.str.contains('sofa', case=False)
df_merchandise.loc[mass_sofa_filter, 'ProductType'] = 'Sofa'

# Repeat for remaining 'armoir' values not captured
armoir_filter = df_merchandise.ProductType.str.contains('armoir', case=False)
df_merchandise.loc[armoir_filter, 'ProductType'] = 'Armoir'

# Repeat for remaining 'mirror' values not captured
mass_mirror_filter = df_merchandise.ProductType.str.contains('mirror', case=False)
df_merchandise.loc[mass_mirror_filter, 'ProductType'] = 'Mirror'

# Repeat for remaining 'picture' values not captured
mass_picture_filter = df_merchandise.ProductType.str.contains('picture', case=False)
df_merchandise.loc[mass_picture_filter, 'ProductType'] = 'Picture'

# Repeat for remaining 'desk' values not captured
mass_desk_filter = df_merchandise.ProductType.str.contains('desk', case=False)
df_merchandise.loc[mass_desk_filter, 'ProductType'] = 'Desk'

# Repeat for remaining 'chest' values not captured
mass_chest_filter = df_merchandise.ProductType.str.contains('chest', case=False)
df_merchandise.loc[mass_chest_filter, 'ProductType'] = 'Chest'

# Repeat for remaining 'console' values not captured
mass_console_filter = df_merchandise.ProductType.str.contains('console', case=False)
df_merchandise.loc[mass_console_filter, 'ProductType'] = 'Console'

# Repeat for remaining 'free' values not captured
mass_free_filter = df_merchandise.ProductType.str.contains('free', case=False)
df_merchandise.loc[mass_free_filter, 'ProductType'] = 'Free'

# Repeat for remaining 'cabinet' values not captured
mass_cabinet_filter = df_merchandise.ProductType.str.contains('cabinet', case=False)
df_merchandise.loc[mass_cabinet_filter, 'ProductType'] = 'Cabinet'

# Repeat for remaining 'book' values not captured
mass_book_filter = df_merchandise.ProductType.str.contains('book', case=False)
df_merchandise.loc[mass_book_filter, 'ProductType'] = 'Book'

# Repeat for remaining 'bench' values not captured
mass_bench_filter = df_merchandise.ProductType.str.contains('bench', case=False)
df_merchandise.loc[mass_bench_filter, 'ProductType'] = 'Bench'

# Repeat for remaining 'stand' values not captured, including 'Nightstand', and consolidate as 'Stands'
mass_stands_filter = df_merchandise.ProductType.str.contains('stand', case=False)
df_merchandise.loc[mass_stands_filter, 'ProductType'] = 'Stands'

# Repeat for remaining 'dresser' values not captured
mass_dresser_filter = df_merchandise.ProductType.str.contains('dresser', case=False)
df_merchandise.loc[mass_dresser_filter, 'ProductType'] = 'Dresser'

# Repeat for remaining 'clock' values not captured
mass_clock_filter = df_merchandise.ProductType.str.contains('clock', case=False)
df_merchandise.loc[mass_clock_filter, 'ProductType'] = 'Clock'

df_merchandise.ProductType.value_counts().head(25)

Accessory    2202
Decor        1622
Chairs       1567
Table        1326
Art          1160
Lighting     1158
Pillow        837
Rug           504
Sofa          333
Mirror        313
Bed           277
Picture       246
Desk          185
Book          179
None          150
Stands        134
Free          122
Cabinet       121
Chest         121
Console       114
Ottoman        98
Bench          69
Dresser        67
Clock          56
Armoir         53
Name: ProductType, dtype: int64

In [42]:
df_merchandise[df_merchandise.groupby('ProductType')['ProductType'].transform('size') < 100].shape

(2636, 20)

Earlier there were 4,386 rows that were to be excluded from the dataset when analyzing top 20 selling items. After cleaning, that value is now only 2,636 rows.

In [43]:
df_merchandise.head(10)

ProductType        ProductDescription  OrigPrice    Price  \
ProductID                                                             
-2           Lighting                      Lamp     147.60   147.60   
-1             Chairs  Wrought Iron W. 5 Chairs     124.75   124.75   
 1               None             Candle Holder       2.00     2.00   
 2               None                     Colum      49.50    49.50   
 3           Lighting             Credenza Lamp      71.10    71.10   
 4             Pillow                    Pillow     120.00   120.00   
 5                Rug                       Rug     269.10   269.10   
 6               None                   3 Balls       7.50     7.50   
 7              Table       Teak Root/Glass Top    4260.00  4260.00   
 8                Art                       Art       5.50     5.50   

           EnterDate TransferDate  TransferPrice TransferInd TransferBy  \
ProductID                                                                 
-2        2018-08-18   2018-08-18         147.60        Sold      Chris   
-1        2018-07-03   2018-07-03         124.75        Sold    Adriana   
 1        2020-05-20   2020-05-20           2.00        Sold       Gina   
 2        2020-07-18   2020-07-18          49.50        Sold       Noel   
 3        2018-08-25   2018-08-25          71.10        Sold    Adriana   
 4        2018-08-30   2018-08-30         120.00        Sold       Noel   
 5        2018-09-28   2018-09-28         269.10        Sold    Adriana   
 6        2018-10-21   2018-10-21           7.50        Sold      Chris   
 7        2015-09-18          NaT           0.00       Store       Noel   
 8        2019-01-02   2019-01-02           5.50        Sold    Adriana   

           CustomerID  ReqPrice  InvoiceNumber  Return MarkdownDate1  \
ProductID                                                              
-2                  1       0.0           8150   False           NaT   
-1                237       0.0           7749   False           NaT   
 1                  1       0.0          13211   False           NaT   
 2                237       0.0          13842   False           NaT   
 3                  1       0.0           8225   False           NaT   
 4                  1       0.0           8262   False           NaT   
 5                  1       0.0           8479   False           NaT   
 6                  1       0.0           8686   False           NaT   
 7                  1       0.0              0    True           NaT   
 8                237       0.0           9212   False           NaT   

          MarkdownDate2 MarkdownDate3  MercCut  ItemCount ExpireDate  \
ProductID                                                              
-2                  NaT           NaT      0.0          0 2018-11-16   
-1                  NaT           NaT      0.5          0 2018-10-01   
 1                  NaT           NaT      0.0          0 2020-08-18   
 2                  NaT           NaT      0.5          0 2020-10-16   
 3                  NaT           NaT      0.0          0 2018-11-23   
 4                  NaT           NaT      0.0          0 2018-11-28   
 5                  NaT           NaT      0.0          0 2018-12-27   
 6                  NaT           NaT      0.0          0 2019-01-19   
 7                  NaT           NaT      0.0          0 2015-12-17   
 8                  NaT           NaT      0.5          0 2019-04-02   

           BuyersFee  
ProductID             
-2               0.0  
-1               0.0  
 1               0.0  
 2               0.0  
 3               0.0  
 4               0.0  
 5               0.0  
 6               0.0  
 7               0.0  
 8               0.0

In [44]:
df_merchandise[df_merchandise.ProductType == 'None'].ProductDescription.value_counts()

* Unknown               38
Fabric                   5
Oklahoma Sign            3
Tieback                  3
Candle Holder            3
Brackets                 3
Corner Elbow             2
Tiebacks                 2
Bracket                  2
Candle                   2
Hooks                    2
Fluer Dilis              1
Grey/olive Loveseat      1
Spode                    1
Bench                    1
Glass                    1
Sconce                   1
Fur                      1
Milk Jug (2)             1
Mantle                   1
Red Pot                  1
Buffet                   1
Round Tieback            1
New Mattress             1
BREAD BOX                1
Barstool                 1
Cow Hide                 1
Green Easel              1
*NEW* 5x7                1
Hutch                    1
                        ..
Ice Blue Chaise          1
Railing                  1
Curtain Holders          1
Box                      1
Queen Mat/box            1
2 Twin Foundations       1
T

In [53]:
none_filter = df_merchandise[(df_merchandise['ProductType'] == 'None') & ((df_merchandise['ProductDescription'].str.contains('dress', case=False)))]
none_filter

ProductType ProductDescription  OrigPrice  Price  EnterDate  \
ProductID                                                               
20921            None            Dresser      431.2  431.2 2018-09-01   

          TransferDate  TransferPrice TransferInd TransferBy  CustomerID  \
ProductID                                                                  
20921       2018-09-01          431.2        Sold    Adriana           1   

           ReqPrice  InvoiceNumber  Return MarkdownDate1 MarkdownDate2  \
ProductID                                                                
20921           0.0           8287   False           NaT           NaT   

          MarkdownDate3  MercCut  ItemCount ExpireDate  BuyersFee  
ProductID                                                          
20921               NaT      0.0          0 2018-11-30        0.0